deployment/terraform/utils.go

```diff
@@ -352,20 +355,114 @@ func getServerURL(output *Output, deploymentConfig *deployment.Config) string {
 	return deploymentConfig.ServerScheme + "://" + url
 }
 
-// GetAWSConfig returns the AWS config, using the profile configured in the
-// deployer if present, and defaulting to the default credential chain otherwise
-func (t *Terraform) GetAWSConfig() (aws.Config, error) {
+// Durations for the expiry of the AWS Role credentials and for its refresh
+// interval, which, out of an abundance of caution, is a bit shorter than half
+// the expiry duration of the role.
+const (
+	// One hour is the maximum allowed by role chaining
+	AWSRoleTokenDuration        = 1 * time.Hour
+	AWSRoleTokenRefreshInterval = 25 * time.Minute
+)
+
+func (t *Terraform) InitCreds() error {
 	regionOpt := awsconfig.WithRegion(t.config.AWSRegion)
+	var opts []func(*awsconfig.LoadOptions) error
+
+	opts = append(opts, regionOpt)
+
+	if t.config.AWSProfile != "" {
+		opts = append(opts, awsconfig.WithSharedConfigProfile(t.config.AWSProfile))
+	}
+
+	cfg, err := awsconfig.LoadDefaultConfig(context.Background(), opts...)
+	if err != nil {
+		return fmt.Errorf("unable to load default config: %w", err)
+	}
+
+	if t.config.AWSRoleARN != "" {
+		// See https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/credentials/stscreds#hdr-Assume_Role
+		// Create the credentials from AssumeRoleProvider to assume the role
+		// identified by AWSRoleARN, and use them in the loaded config
+		stsSvc := sts.NewFromConfig(cfg)
+		creds := stscreds.NewAssumeRoleProvider(stsSvc, t.config.AWSRoleARN)
+		cfg.Credentials = aws.NewCredentialsCache(creds)
+	}
+
+	t.awsCfgMut.Lock()
+	t.awsCfg = cfg
+	t.awsCfgMut.Unlock()
+
+	if t.config.AWSRoleARN != "" {
+		// Refresh the credentials once so that the time of the interval below
+		// starts now
+		if err := t.refreshAWSCredentialsFromRole(); err != nil {
+			return fmt.Errorf("unable to refresh token: %w", err)
+		}
+
+		// Start a goroutine refreshing the credentials every AWSRoleTokenRefreshInterval.
+		// If it fails, exit the goroutine, since it needs the previous credentials to be
+		// valid for the refresh logic to work.
+		go func() {
```

Comment:
Do we need to bother about shutting down the goroutine during exit? Or just letting it die is good enough?

Reply:
I thought about this, but there is really no explicit exit. When we create the `Terraform` struct, we can use it for `Create`, or for `Destroy`, or for inspecting the output, but there is no explicit signal of "we're done with this", so I just decided to let it die.